### PR TITLE
Updated Decrypting Realm Databases in 0x05d - Testing Data Storage and 0x06d-Testing-Data-Storage 

### LINKED ISSUE #1951: Decrypting Realm Databases
**Proposal to add to a section**
Adding to section [0x06-d - Realm Databases](https://github.com/OWASP/owasp-mstg/blob/master/Document/0x06d-Testing-Data-Storage.md#realm-databases)

### Decrypting Realm Databases
However Realm Databases have a notable issue when using the encrypted method, The encryption key is also the decryption key. If a user is able to obtain the key that is used to encryption they are able to decrypt the database and view the contents contained in it. This is trivial to do with a simple Frida script for both iOS & Android.

Resources for Android: [Enter The Realm](https://x64.sh/mobile/security/2020/12/21/Enter-The-Realm/)

### PR DESCRIPTION
**Pull request by R3zk0n**
Thank you for submitting a Pull Request to the OWASP MASTG. Please make sure that:

- [X] Your contribution is written in the 2nd person (e.g. you)
- [X] Your contribution is written in an active present form for as much as possible.
- [X] You have made sure that the reference section is up to date (e.g. please add sources you have used, make sure that the references to MITRE/MASVS/etc. are up to date)
- [X] Your contribution has proper formatted markdown and/or code
- [X] Any references to website have been formatted as [TEXT](URL “NAME”)
- [X] You verified/tested the effectiveness of your contribution (e.g.: is the code really an effective remediation? Please verify it works!)

If your PR is related to an issue. Please end your PR test with the following line:
This PR closes #1951

Long delay and update, added both iOS and Android examples

### CODE REVIEW

**Review by R3zk0n — COMMENTED** [review]
Updated Frida Script

File: Document/0x05d-Testing-Data-Storage.md [hunk 1]
```diff
@@ -178,6 +178,68 @@ Realm realm = Realm.getInstance(config);
 
 If the database _is not_ encrypted, you should be able to obtain the data. If the database _is_ encrypted, determine whether the key is hard-coded in the source or resources and whether it is stored unprotected in shared preferences or some other location.
 
+However its quite important to be aware that if the database _is_ encrypted, its often possible to obtain the decryption key at runtime. This is because the encryption and decryption keys are identical and are invoked at runtime to facilitate access to the Realm file. The frida script below is demonstrating how to intercept the specific Realm key utilized by the Realm database, allowing for the decryption of encrypted database.
```
Comment (cpholguera) [inline Document/0x05d-Testing-Data-Storage.md:181]:
```suggestion
Accessing the data depends on the encryption: unencrypted databases are easily accessible, while encrypted ones require investigating how the key is managed-whether it's hardcoded or stored unencrypted in an insecure location such as shared preferences, or securely within the platform's KeyStore (which is best practice).

However, if an attacker has sufficient access to the device (e.g. root access) or can repackage the app, they can still retrieve encryption keys at runtime using tools like Frida. The following Frida script demonstrates how to intercept the Realm encryption key and access the contents of the encrypted database.
```

File: Document/0x06d-Testing-Data-Storage.md [hunk 2]
```diff
@@ -94,6 +94,48 @@ do {
   fatalError("Error opening realm: \(error)")
 }
 ```
+One security concern that warrants attention involves the potential interception or compromise of the encryption key when accessing the Realm database. This arises due to the necessity of supplying the decryption key at runtime, which introduces a window to capture the decryption key. The frida script demonstrated below targets the RLMRealmConfiguration class within the Realm database framework, leveraging its functionality to extract the decryption key. By hooking into this class, the script retrieves the key directly from memory, converting it into a hexadecimal string which then can be used to decrypt the database.
```
Comment (cpholguera) [inline Document/0x06d-Testing-Data-Storage.md:97]:
```suggestion
Accessing the data depends on the encryption: unencrypted databases are easily accessible, while encrypted ones require investigating how the key is managed-whether it's hardcoded or stored unencrypted in an insecure location such as shared preferences, or securely within the platform's KeyChain (which is best practice).

However, if an attacker has sufficient access to the device (e.g. jailbroken device) or can repackage the app, they can still retrieve encryption keys at runtime using tools like Frida. The following Frida script demonstrates how to intercept the Realm encryption key and access the contents of the encrypted database.
```

**Review by cpholguera — CHANGES_REQUESTED** [review]

**Comment by cpholguera** [comment]
Maybe you could include a final short paragraph in both files indicating that the DB can be then extracted with `adb pull` and viewed with https://github.com/realm/realm-studio could you please try if that works for you?

Something like

```
To access the decrypted data, you can pull the database using `adb pull`, then load it into [Realm Studio](https://docs.realm.io/sync/realm-studio#browsing-local-realms) and insert the key you previously obtained.
```

File: Document/0x05d-Testing-Data-Storage.md (hunk 1 above)
Comment (sushi2k) [reply Document/0x05d-Testing-Data-Storage.md:181]: @cpholguera  slight changes to your suggestion in the 1st paragraph (checked with DeepL):

```suggestion
Access to the data depends on the encryption: unencrypted databases are easily accessible, while encrypted ones require investigation into how the key is managed - whether it's hard-coded or stored unencrypted in an insecure location such as shared preferences, or securely in the platform's KeyStore (which is best practice).

However, if an attacker has sufficient access to the device (e.g. root access) or can repackage the app, they can still retrieve encryption keys at runtime using tools like Frida. The following Frida script demonstrates how to intercept the Realm encryption key and access the contents of the encrypted database.
```

File: Document/0x06d-Testing-Data-Storage.md (hunk 2 above)
Comment (sushi2k) [reply Document/0x06d-Testing-Data-Storage.md:97]: see comment above.

File: Document/0x05d-Testing-Data-Storage.md [hunk 3]
```diff
@@ -176,7 +176,70 @@ Realm realm = Realm.getInstance(config);
 
 ```
```
Comment (cpholguera) [inline Document/0x05d-Testing-Data-Storage.md:178]:
```suggestion
```

**Review by cpholguera — APPROVED** [review]
Thanks for the additions @R3zk0n!

File: Document/0x06d-Testing-Data-Storage.md [hunk 4]
```diff
@@ -95,6 +95,51 @@ do {
 }
 ```
 
+Access to the data depends on the encryption: unencrypted databases are easily accessible, while encrypted ones require investigation into how the key is managed - whether it's hardcoded or stored unencrypted in an insecure location such as shared preferences, or securely in the platform's KeyStore (which is best practice).
+
```
Comment (cpholguera) [inline Document/0x06d-Testing-Data-Storage.md:99]:
```suggestion

```

**Pull request merged**